File: lint-arwaky/archive/taxonomy/file_system_error.py
"""file_system_error — File system domain error types."""

from pydantic import BaseModel, ConfigDict

from .error_code_vo import ErrorCode
from .error_value_vo import (
    ErrorMessage,
    Cause,
)
from .file_path_vo import FilePath
from .job_action_vo import ActionName


class FileSystemError(BaseModel):
    """General file system operation failure."""

    model_config = ConfigDict(frozen=True)

    path: FilePath
    message: ErrorMessage
    operation: ActionName  # e.g., 'read', 'write', 'delete', 'walk'
    error_code: ErrorCode | None = None
    cause: Cause | None = None

    def __str__(self):
        code = f" [{self.error_code}]" if self.error_code else ""
        return f"FS Error during {self.operation} on {self.path}{code}: {self.message}"


class PathNotFoundError(FileSystemError):
    """Target path does not exist on disk."""

    def __str__(self):
        return f"Path not found: {self.path} ({self.message})"


class AccessDeniedError(FileSystemError):
    """Permission denied for the requested operation."""

    def __str__(self):
        return f"Access denied: {self.path} ({self.message})"

File: lint-arwaky/src-rust/taxonomy/file_system_error.rs
use serde::{Serialize, Deserialize};
use super::*;

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct FileSystemError {
    pub path: FilePath,
    pub message: ErrorMessage,
    pub operation: ActionName,
    #[serde(default)]
    pub error_code: Option<ErrorCode>,
    #[serde(default)]
    pub cause: Option<Cause>,
}

impl FileSystemError {
    pub fn new(path: FilePath, message: ErrorMessage, operation: ActionName) -> Self {
        Self { path, message, operation, error_code: None, cause: None }
    }
}

impl std::fmt::Display for FileSystemError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let code = self.error_code.as_ref().map(|c| format!(" [{}]", c)).unwrap_or_default();
        write!(f, "FS Error during {} on {}{}: {}", self.operation, self.path, code, self.message)
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct PathNotFoundError {
    #[serde(flatten)]
    pub base: FileSystemError,
}

impl PathNotFoundError {
    pub fn new(path: FilePath, message: ErrorMessage) -> Self {
        Self { base: FileSystemError::new(path, message, ActionName::new("read")) }
    }
}

impl std::fmt::Display for PathNotFoundError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "Path not found: {} ({})", self.base.path, self.base.message)
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct AccessDeniedError {
    #[serde(flatten)]
    pub base: FileSystemError,
}

impl AccessDeniedError {
    pub fn new(path: FilePath, message: ErrorMessage) -> Self {
        Self { base: FileSystemError::new(path, message, ActionName::new("access")) }
    }
}

impl std::fmt::Display for AccessDeniedError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "Access denied: {} ({})", self.base.path, self.base.message)
    }
}
